### PR TITLE
Add support for AWS Network Load Balancers

### PR DESCRIPTION
**Pull request by csrwng**
- Updates the Makefile to fix `make generate`
- Adds a LoadBalancerType field to the provider config
- Updates the actuator to work with the new network Load Balancers

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -57,7 +57,7 @@ test: generate-mocks unit
 gendeepcopy:
 	go build -o $$GOPATH/bin/deepcopy-gen sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/code-generator/cmd/deepcopy-gen
 	deepcopy-gen \
```
Comment (enxebre) [inline Makefile]: thanks! we should actually rely on go generate for this now https://github.com/openshift/cluster-api-provider-aws/blob/master/pkg/apis/apis.go#L18 cc @spangenberg

Comment (csrwng) [reply Makefile]: @enxebre thx, fixed.

File: pkg/apis/awsproviderconfig/v1alpha1/awsmachineproviderconfig_types.go [hunk 2]
```diff
@@ -135,7 +135,11 @@ type AWSMachineProviderConfig struct {
 
 	// LoadBalancerNames is the names of the load balancers to which the new instance
 	// should be added once it is created.
-	LoadBalancerNames []string `json:"loadBalancerIds"`
+	LoadBalancerNames []string `json:"loadBalancerNames"`
```
Comment (enxebre) [inline pkg/apis/awsproviderconfig/v1alpha1/awsmachineproviderconfig_types.go]: Would it make sense for this to be a common struct? e.g:
```
type loadBalancers struct {
	Names
	type
}
```

File: hack/go-lint.sh [hunk 3]
```diff
@@ -7,8 +7,8 @@ if [ "$IS_CONTAINER" != "" ]; then
 else
   docker run --rm \
     --env IS_CONTAINER=TRUE \
```
Comment (enxebre) [inline hack/go-lint.sh]: thanks!

File: pkg/apis/awsproviderconfig/v1alpha1/awsmachineproviderconfig_types.go (hunk 2 above)
Comment (csrwng) [reply pkg/apis/awsproviderconfig/v1alpha1/awsmachineproviderconfig_types.go]: Added a load balancer ref type  that has both name and type, ptal

**Comment by csrwng** [comment]
@frobware can you please approve this if it looks good to you? thx!

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 4]
```diff
@@ -603,8 +605,17 @@ func (a *Actuator) getMachineInstances(cluster *clusterv1.Cluster, machine *clus
...
 		return nil
 	}
+	if providerConfig.LoadBalancerType == providerconfigv1.NetworkLoadBalancer {
+		return a.registerWithNetworkLoadBalancers(client, providerConfig, instance)
+	}
+	return a.registerWithClassicLoadBalancers(client, providerConfig, instance)
+}
+
+func (a *Actuator) registerWithClassicLoadBalancers(client awsclient.Client, providerConfig *providerconfigv1.AWSMachineProviderConfig, instance *ec2.Instance) error {
+	glog.V(4).Infof("Updating classic load balancer registration for %s\n", *instance.InstanceId)
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator.go]: Drop the `\n` - 
```
// Infof logs to the INFO log.
// Arguments are handled in the manner of fmt.Printf; a newline is appended if missing.
func Infof(format string, args ...interface{}) {
	logging.printf(infoLog, format, args...)
}
```

Also, same nit :) for `%s` versus `%q`.

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 5]
```diff
@@ -619,7 +630,67 @@ func (a *Actuator) UpdateLoadBalancers(client awsclient.Client, providerConfig *
 	}
 
 	if len(errs) > 0 {
-		return fmt.Errorf("failed to register instances with elbs: %v", errs)
+		return fmt.Errorf("failed to register instances with classic ELBs: %v", errs)
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator.go]: Observation: `...%v", errs` will format as `[foo bar baz]` - presumably that's OK?

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 6]
```diff
@@ -619,7 +630,67 @@ func (a *Actuator) UpdateLoadBalancers(client awsclient.Client, providerConfig *
...
+	}
+	return nil
+}
+
+func (a *Actuator) registerWithNetworkLoadBalancers(client awsclient.Client, providerConfig *providerconfigv1.AWSMachineProviderConfig, instance *ec2.Instance) error {
+	glog.V(4).Infof("Updating network load balancer registration for %s\n", *instance.InstanceId)
+	lbNames := []*string{}
```
Comment (frobware) [inline pkg/cloud/aws/actuators/machine/actuator.go]: nit: `lbNames := make([]*string, len(providerConfig.LoadBalancerNames)` - as we know up front the length.

**Comment by frobware** [comment]
> @frobware can you please approve this if it looks good to you? thx!

Apologies for the latency. Dropped some comments, but all are nits really.

**Comment by csrwng** [comment]
@frobware thank you much, comments addressed

File: Makefile [hunk 7]
```diff
@@ -48,23 +48,11 @@ vendor:
 	patch -p1 < 0002-Sort-machines-before-syncing.patch
 
 .PHONY: generate
-generate: gendeepcopy generate-mocks
+generate: 
+	go generate ./pkg/... ./cmd/...
```
Comment (ingvagabund) [inline Makefile]: So `go generate` now can run the `deepcopy-gen` binary? Resp. why was `gendeepcopy` removed from the Makefile?

Comment (enxebre) [reply Makefile]: See https://github.com/openshift/cluster-api-provider-aws/blob/master/pkg/apis/apis.go#L18
and deepcopy-gen is required by Gopkg. 
This is all driven by kubebuilder generation

**Pull request merged**